gdnative-core/src/class.rs

```diff
@@ -60,71 +69,148 @@ pub trait NativeClassMethods: NativeClass {
     fn register(builder: &crate::init::ClassBuilder<Self>);
 }
 
-/// A reference to a rust native script.
-pub struct NativeRef<T: NativeClass> {
-    this: *mut sys::godot_object,
-    _marker: PhantomData<T>,
+/// A reference to a GodotObject with a rust NativeClass attached.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Debug)]
+pub struct Instance<T: NativeClass> {
+    this: T::Base,
+    script: T::UserData,
 }
 
-impl<T: NativeClass> NativeRef<T> {
-    /// Try to cast into a godot object reference.
-    pub fn cast<O>(&self) -> Option<O>
+impl<T: NativeClass> Instance<T> {
+    /// Creates a `T::Base` with the script `T` attached. Both `T::Base` and `T` must have zero
+    /// argument constructors.
+    /// 
+    /// Must be called after the library is initialized.
+    pub fn new() -> Self
     where
-        O: GodotObject,
+        T::Base: Instanciable,
     {
-        object::godot_cast::<O>(self.this)
-    }
-
-    /// Creates a new reference to the same object.
-    pub fn new_ref(&self) -> Self {
         unsafe {
-            object::add_ref(self.this);
+            let gd_api = get_api();
+            
```

Comment:
I think a comment here explaining that all string literals here have to be NUL terminated instead of using `CString`.

Reply:
`CString` is owned and `&CStr`'s constructor is not `const`. Both has runtime cost. A comment is added there.